media/Jsontracer/processData.js

```diff
@@ -0,0 +1,132 @@
+//    * Redistributions in binary form must reproduce the above
+// copyright notice, this list of conditions and the following disclaimer
+// in the documentation and/or other materials provided with the
+// distribution.
+//    * Neither the name of Google Inc. nor the names of its
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+// "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+// LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+// A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+// OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+// LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+// OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file referenced the result of https://github.com/catapult-project/catapult/tree/444aba89e1c30edf348c611a9df79e2376178ba8/tracing
+
+// TODO import renderDashboard from "./dashboard.js";
+import getColorList from "./colorList.js";
+
+const setData = document.querySelector(".set-data");
+
+export default function openFileSelector() {
+  const input = document.createElement("input");
+  input.type = "file";
+  input.accept = "text/plain";
+  input.onchange = (event) => {
+    setFileName(event.target.files[0].name);
+    processFile(event.target.files[0]);
+  };
+  input.click();
+}
+
+function setFileName(name) {
+  const fileName = document.querySelector(".file-name");
+  fileName.innerText = name;
+}
+
+function processFile(file) {
+  const reader = new FileReader();
+  reader.onload = () => {
+    const data = JSON.parse(reader.result);
+    processData(data.traceEvents);
+
+    // set data to DOM
+    setData.dataset["displayTimeUnit"] = data.displayTimeUnit;
+  };
+  reader.readAsText(file, "euc-kr");
+}
+
+function processData(data) {
+  const processedData = {};
+  const backgroundColor = {};
+  const utility = {};
+  const colorList = getColorList();
+  const colorLen = colorList.length;
+  let endTime = 0;
+  let colorIdx = 0;
+
+  data.forEach((ele, idx) => {
+    // exception handling
+    if (!ele.name) {
+      return;
+    }
+
+    // set data object
+    if (processedData[ele.pid] === undefined) {
+      processedData[ele.pid] = {};
+    }
+    if (processedData[ele.pid][ele.tid] === undefined) {
+      processedData[ele.pid][ele.tid] = [];
+    }
+
+    // set backgroud-color
+    if (!backgroundColor[ele.name]) {
+      backgroundColor[ele.name] = colorList[colorIdx];
+      colorIdx += 1;
+      colorIdx %= colorLen;
```

Comment:
From above comment, I mean this one!

```suggestion
      colorIdx %= colorList.length;
```

Reply:
I got it! I will change this.
But I'm curious. If we modify it with `colorList.length`, wouldn't it be slower than the existing code by calculating the length every time?
Is there a reason why you want to fix it?

Reply:
> wouldn't it be slower than the existing code by calculating the length every time?

 `Array` has a metadata which includes length information too.
Therefore, time complexity of `Array.length` is O(1) which is exactly same with current implementation :)

> Is there a reason why you want to fix it?

Hmm... technically, it is not mandatory.
But usually, we recommends to define the variable according to the scope for preventing mistake and better readability!

Reply:
Oh, I see. Thank you so much! I'll fix it.